File: lib/lib.error.php
<?php declare(strict_types=1);
/**
 * Error handling functions
 *
 * Part of the DOMjudge Programming Contest Jury System and licensed
 * under the GNU GPL. See README and COPYING for details.
 */
if (! defined('SCRIPT_ID')) {
    define('SCRIPT_ID', basename($_SERVER['PHP_SELF'], '.php'));
}

// Default verbosity and loglevels:
$verbose  = LOG_NOTICE;
$loglevel = LOG_DEBUG;

// Open standard error:
if (!defined('STDERR')) {
    define('STDERR', fopen('php://stderr', 'w'));
}

// Open log file:
if (defined('LOGFILE')) {
    if ($fp = @fopen(LOGFILE, 'a')) {
        define('STDLOG', $fp);
    } else {
        fwrite(STDERR, "Cannot open log file " . LOGFILE .
            " for appending; continuing without logging.\n");
    }
    unset($fp);
}

// Open syslog connection:
if (defined('SYSLOG')) {
    if (!openlog(SCRIPT_ID, LOG_NDELAY | LOG_PID | LOG_CONS, SYSLOG)) {
        error("cannot open syslog");
    }
}

/**
 * Log a message $string on the loglevel $msglevel.
 * Prepends a timestamp and log to the logfile.
 * If this is the web interface: write to the screen with the right CSS class.
 * If this is the command line: write to Standard Error.
 */
function logmsg(int $msglevel, string $string)
{
    global $verbose, $loglevel;

    // Trim $string to reasonable length
    $string = substr($string, 0, 10000);

    $msec = sprintf("%03d", (int)(explode(' ', microtime())[0]*1000));
    $stamp = "[" . date('M d H:i:s') . ".$msec] " . SCRIPT_ID .
        (function_exists('posix_getpid') ? "[" . posix_getpid() . "]" : "") .
        ": ";

    if ($msglevel <= $verbose) {
        fwrite(STDERR, $stamp . $string . "\n");
        fflush(STDERR);
    }
    if ($msglevel <= $loglevel) {
        if (defined('STDLOG')) {
            fwrite(STDLOG, $stamp . $string . "\n");
            fflush(STDLOG);
        }
        if (defined('SYSLOG')) {
            // Remove the ANSI escape characters,
            // When logging direct to the terminal the checkmark should be green, this
            // does not work in log so remove this specific control char.
            // This removes both the start and the reset of the color.
            syslog($msglevel, preg_replace('# ?\\x1b\[[0-9];?[0-9]*m#', '', $string) . "\n");
        }
    }
}

/**
 * Log an error at level LOG_ERROR and exit with exitcode 1.
 */
function error(string $string)
{
    logmsg(LOG_ERR, "error: $string");
    exit(1);
}

/**
 * Log a warning at level LOG_WARNING.
 */
function warning(string $string)
{
    logmsg(LOG_WARNING, "warning: $string");
}

/**
 * Write debugging output using var_dump(). Uses <pre> in web context.
 */
function debug()
{
    if (DEBUG===0) {
        return;
    }

    call_user_func_array('var_dump', func_get_args());
}

/**
 * Handle exceptions by calling error().
 */
function exception_handler(Throwable $e)
{
    error($e->getMessage() . ( (defined('DEBUG') && DEBUG) ? " in " . $e->getFile() . " line " . $e->getLine() : ''));
}

set_exception_handler('exception_handler');

File: lib/lib.error.cc
/*
 * Error handling and logging functions
 *
 * Part of the DOMjudge Programming Contest Jury System and licensed
 * under the GNU GPL. See README and COPYING for details.
 */

#include "config.h"

#include "lib.error.hpp"

#include <cstdlib>
#include <cstring>
#include <cstdarg>
#include <cstdio>
#include <vector>
#include <string>
#include <iostream>
#include <fstream>
#include <unistd.h>
#include <chrono>

/* Use program name in syslogging if defined */
#ifndef PROGRAM
#define PROGRAM NULL
#endif

const int exit_failure = -1;

/* Variables defining logmessages verbosity to stderr/logfile */
int  verbose      = LOG_NOTICE;
int  loglevel     = LOG_DEBUG;

/* Variables for tracking logging facilities */
static std::ofstream stdlog;
int  syslog_open  = 0;

/* Core logging implementation that accepts a fully formatted string */
void logmsg_str(int msglevel, const std::string& mesg)
{
	std::string buffer;
	char *str, *endptr;
	int syslog_fac;

	/* Try to open logfile if it is defined */
#ifdef LOGFILE
	if (!stdlog.is_open()) {
		stdlog.open(LOGFILE, std::ios::out | std::ios::app);
	}
#endif

	/* Try to open syslog if it is defined */
	if ( ! syslog_open && (str=getenv("DJ_SYSLOG"))!=NULL ) {
		syslog_fac = strtol(str,&endptr,10);
		if ( *endptr==0 ) {
			openlog(PROGRAM, LOG_NDELAY | LOG_PID, syslog_fac);
			syslog_open = 1;
		}
	}

	auto now = std::chrono::system_clock::now();
	auto millis = std::chrono::duration_cast<std::chrono::milliseconds>(now.time_since_epoch()) % 1000;
	std::string timestring = std::format("{:%b %d %H:%M:%S}.{:03d}",
			std::chrono::floor<std::chrono::seconds>(now), millis.count());

	// Construct the format string: "[time] progname[pid]: message\n"
	buffer += "[" + timestring + "] ";
	buffer += progname;
	buffer += "[" + std::to_string(getpid()) + "]: ";
	buffer += mesg;

	if ( msglevel<=verbose ) {
		std::cerr << buffer << std::endl;
	}
	if ( msglevel<=loglevel && stdlog.is_open() ) {
		stdlog << buffer << std::endl;
	}

	if ( msglevel<=loglevel && syslog_open ) {
		syslog(msglevel, "%s", mesg.c_str());
	}
}
